src/main/java/spoon/support/visitor/equals/EqualsVisitor.java

```diff
@@ -107,6 +114,35 @@ public void biScan(CtRole role, CtElement element, CtElement other) {
 		}
 	}
 
+	private boolean haveDifferentDeclaringInnerClasses(CtElement element, CtElement other) {
+		if (element instanceof CtReference || other instanceof CtReference) {
+			return false;
+		}
+		CtClass innerClassParent1;
+		CtClass innerClassParent2;
+		try {
+			innerClassParent1 = element.getParent(new TypeFilter<>(CtClass.class));
+			innerClassParent2 = other.getParent(new TypeFilter<>(CtClass.class));
+		} catch (ParentNotInitializedException e) {
+			return false;
+		}
+		if (innerClassParent1 == null || innerClassParent2 == null
+				|| (innerClassParent1.isTopLevel() && innerClassParent2.isTopLevel())) {
+			return false;
+		}
+		if ((!innerClassParent1.isTopLevel() && innerClassParent2.isTopLevel())
+				|| (innerClassParent1.isTopLevel() && !innerClassParent2.isTopLevel())) {
+			return true;
+		}
+		int lastDot1 = innerClassParent1.getQualifiedName().lastIndexOf('.');
+		int lastDot2 = innerClassParent2.getQualifiedName().lastIndexOf('.');
+
+		// remove package name
+		String name1 = innerClassParent1.getQualifiedName().substring(lastDot1 + 1);
```

Comment:
maybe add a comment for the magic +1 addition?

Reply:
I expanded the comment to make the +1 more clear:
```java
// remove package name (everything prior to and including the last dot)
```